### ISSUE TITLE
How to navigate in a redux-action ?

### ISSUE THREAD

**Issue by taipa-ibl** [react-community/react-navigation#840]
I using react-native-router-flux now, it is easy to navigate in a Redux-Action, but how to do it in react-nagivation ?

Thank you.

**Comment by andrerfneves** [comment]
+1

**Comment by daose** [comment]
From https://github.com/daose/react-native-template/issues/1, https://github.com/react-community/react-navigation/issues/153

You can `import { NavigationActions } from 'react-navigation'`, and then if you have `dispatch`, you can do

```javascript
dispatch(NavigationActions.navigate({ routeName: 'Route' });
```

So if you're in a saga

```javascript
yield put(NavigationActions.navigate({ routeName: 'Route' }); // import put from redux-saga/effects
```

There's a lot more you can do with it as well, you can see the available actions at [NavigationActions.js](https://github.com/react-community/react-navigation/blob/master/src/NavigationActions.js)

**Comment by andrerfneves** [comment]
^ just spent some time fixing it in my application. For those that run into this issue, you must also hold your application state inside your Redux store in order to dispatch NavigationActions from your action creators.

Cheers @daose!

**Comment by tmaly1980** [comment]
@andrerfneves , good to know but not enough info to actually accomplish that. What I ended up doing was adding this to my AppNavigator (parent of the root container):

in AppNavigator.js:

```
import React, { Component } from 'react';
import { addNavigationHelpers } from 'react-navigation';
import { bindActionCreators } from 'redux';
import { ActionCreators } from './actions';
import { RootContainer } from './components/router';
import { connect } from 'react-redux';

import reducer from './reducers';

const store = createStore(reducer); // possibly apply some middleware/initial state

class AppNavigator extends Component {
  render() {
    return (
      <Provider store={store}>
        <RootContainer 
          navigation={addNavigationHelpers({ 
            dispatch: this.props.dispatch,
            state: this.props.navigation,
          })} 
        />
      </Provider>
    );
  }

const bindAction = dispatch => {
    return Object.assign({dispatch: dispatch}, bindActionCreators(ActionCreators, dispatch)); 
    // add dispatch itself to props, so available for addNavigationHelpers
};

const mapStateToProps = state => ({
  navigation: state.navigation, // needed for addNavigationHelpers
});

export default connect(mapStateToProps, bindAction)(AppNavigator);
```

In components/router.js:

```
import { TabNavigator, StackNavigator, TabView } from 'react-navigation';
import Component1 from './Component1'; // 
import Component2 from './Component2';

export const RootContainer = TabNavigator({
  Component1: { 
    screen: Component1
  },
  Component2: {
     screen: Component2
  }
}
```

In actions/index.js: 

```
import * as SomeActions from './someActions';

export const ActionCreators = Object.assign({},
  someActions
  // whatever else
);
```

In actions/someActions.js:

```
// You can call this from any reduxed component as this.props.goToComponent1(params)
export function goToComponent1(params) 
{
  return (dispatch, getState) => {
    dispatch(NavigationActions.navigate({ routeName: 'Component1', params: params }));
  };
}
```

In reducers/index.js:

```
import { RootContainer } from '../components/router'; // Object has access to router

const navigation = (state,action) => {
    const newState = RootContainer.router.getStateForAction(action, state);
    return newState || state;
}

export default combineReducers({
    navigation // this needs to match 'this.props.navigation' given to addNavigationHelpers
});
```

I may be missing some pieces, apologies if I do (edited for simplicity). Everything I found out there was less clear and this took me several hours to come up with.

The docs REALLY should have a more robust redux info page, I found it very much lacking.

**Comment by RWOverdijk** [comment]
A lot of people are coming over from react-native-router-flux now, it might be useful to have more in-depth explanation of how to set this up (or a no-nonsense plugin that just makes this work) so they feel right at home (including me).

I'm willing to contribute to that, as soon as I figure out the best way to do so.

TL;DR; 👍

**Comment by ahmedkamohamed** [comment]
This is my effort to extend the github notetaker app done by Tyler Mcginnis. I updated the code to support latest versions of React and React Native, I also extended the app to use react-navigation which the latest navigation library supporting both iOS and Android. I added Redux to manage the store and used Redux-Sagas to handle asynchronous fetch of data. I would appreciate it if you check it out and give it thumbs up if its helpful. I will continue extending to support more features.
https://github.com/ahmedkamohamed/react-native-navigation-redux-sagas

**Comment by manhduc2610** [comment]
@tmaly1980  so good, i do it but i get error "Could not find "store" in either the context or props of "Connect(Main)""
my code : 
import React, { Component } from 'react';
import Home from './index.js';
import {Navigator} from '../navigation/index.js';
import store from '../redux/store.js';
import { Provider } from "react-redux";
import { addNavigationHelpers } from 'react-navigation';
import {ActionCreators} from '../redux/actions/index.js';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';


const bindAction = dispatch => {
  return Object.assign({dispatch: dispatch}, bindActionCreators(ActionCreators, dispatch)); 
};

const mapStateToProps = state => ({
  navigation: state.navigation, 
});
console.log(11214,store)
class Main extends Component {
  render() {
    
    return (
	 		<Provider store={store}>
      	<Navigator 
        navigation={addNavigationHelpers({
          dispatch: this.props.dispatch,
          state: this.props.navigation,
        })}
        />
  		</Provider>
    );
  }
}

export default connect(mapStateToProps,bindAction)(Main)

**Comment by ahmedkamohamed** [comment]
Are you sure you have the correct path to the store component?

**Comment by manhduc2610** [comment]
yes, i console.log(store) , my index.js in reducers:
import { combineReducers } from "redux"
import {Navigator} from "../../navigation/index.js"

import search from "./searchReducer.js"
import home from "./home.js"

const navigation =(state,action) => {
	const newState = Navigator.router.getStateForAction(action,state);
	return newState || state
}

export default combineReducers({
  search,
  home,
  navigation
})

**Comment by impuulse** [comment]
http://joxi.ru/ZrJdn1NT1wGEP2  after this in actions
```
dispatch(NavigationActions.navigate({ routeName: 'Route' });
```
but at this case all okey:
```
<TouchableOpacity onPress={() => this.props.navigation.navigate('PastingList')}>
```
plz help

**Comment by spencercarli** [comment]
Hi! In an effort to get the hundreds of issues on this repo under control I'm closing issues tagged as questions. Please don't take this personally - it's simply something we need to do to get the issues to a manageable point. If you still have a question I encourage you to re-read the docs, ask on StackOverflow, or ask on the #react-navigation Reactiflux channel. Thank you!

**Comment by adrienjt** [comment]
My [comment](https://github.com/react-community/react-navigation/issues/1439#issuecomment-340293063) in issue #1439 offers a simple solution in case react-navigation is NOT integrated with Redux (see also the proposed service solution in that issue).

**Comment by chefarchitekt** [comment]
@adrienjt Thanks for your solution. Just brilliance.

**Comment by franko19** [comment]
@adrienjt You have no idea how much I love you right now

**Comment by vinipachecov** [comment]
Could not understand the solution exposed here. Why not send the navigation prop from the screen into the action and just navigate to the scene you want?

```
export const testAction = ({ navigation }) => {
  navigation.navigate('pathOfMyScene');
  return { type: 'something' };
};

```

**Comment by mvanroon** [comment]
@vinipachecov because that means that you'll have to pass on the navigation prop every time you call an action

**Comment by kaiyisg** [comment]
For those who are still looking, you might need to do [this](https://reactnavigation.org/docs/redux-integration.html) to make sure that react-navigation is in redux, then you can dispatch actions from redux to control navigation

**Comment by hlbp** [comment]
@kaiyisg do you know how can i pass params through react-navigation with redux? i mean 
in my reducer i have 
`case LOGGED_IN:
      //console.log('Entrada al caso LOGGED_IN del NavReducer');
      nextState = AppNavigator.router.getStateForAction(
        NavigationActions.navigate({ routeName: 'MainAppView', params: {'myParam': 'Helmer'} }),
        state,
      );`
then i could extract props in the view im passing the params with that line of code
 `    const { params } = this.props.navigation.state;`
but the params are not been passed. do you have any idea? thanks!

**Comment by 09allude** [comment]
@hlbp  Hi, do you know how pass params now, sir?

**Comment by kakkarott** [comment]
@daose 

> `yield put(NavigationActions.navigate({ routeName: 'Route' });`

Above line in saga results in nothing. Does anything extra needs to be taken care?

**Comment by ruci06** [comment]
@hlbp 
@09allude 
`
NavigationActions.navigate({ routeName: 'ROUTE_NAME', params: { myParam : 'PARAM' }})
`

then
`
this.props.navigation.getParam('myParam')
`
return 'PARAM'

**Comment by souzamarcos** [comment]
Following the tutorial by React Navigation (https://reactnavigation.org/docs/en/navigating-without-navigation-prop.html#docsNav)

"The presented approach is useful in situations when you want to trigger a navigation action from places where you do not have access to the navigation prop, or if you're looking for an alternative to using the navigation prop."

**Comment by souzamarcos** [comment]
Follow the tutorial by React Navigation (https://reactnavigation.org/docs/en/navigating-without-navigation-prop.html#docsNav)

"The presented approach is useful in situations when you want to trigger a navigation action from places where you do not have access to the navigation prop, or if you're looking for an alternative to using the navigation prop."

Create the NavigationService.js:
```
import { NavigationActions } from 'react-navigation';

let _navigator;

function setTopLevelNavigator(navigatorRef) {
  _navigator = navigatorRef;
}

function navigate(routeName, params) {
  _navigator.dispatch(
    NavigationActions.navigate({
      routeName,
      params,
    })
  );
}

// add other navigation functions that you need and export them

export default {
  navigate,
  setTopLevelNavigator,
};
```
"You can get access to a navigator through a ref and pass it to the NavigationService which we will later use to navigate. Use this only with the top-level (root) navigator of your app."

App.js:
```
import NavigationService from './NavigationService';

const TopLevelNavigator = createStackNavigator({ /* ... */ })

class App extends React.Component {
  // ...

  render() {
    return (
      <TopLevelNavigator
        ref={navigatorRef => {
          NavigationService.setTopLevelNavigator(navigatorRef);
        }}
      />
    );
  }
}
```

In you code import the NavigationService
```
import NavigationService from 'path-to-NavigationService.js';
```
and navigate like this example:
```
NavigationService.navigate('Screen2', { param1: 'a', param2: 'b'  });
```

**Comment by sregg** [comment]
@souzamarcos's answers works well until you want to use `push()` instead of `navigate()`.
`NavigationActions` doesn't have `push()`.
Is there another way of doing that?

**Comment by erinha** [comment]
@souzamarcos I've come across that example and it works great, except I'm having trouble retrieving the `params` passed to `navigate`.

Do you know how to retrieve `param1` passed in the `NavigationService.navigate('Screen2', { param1: 'a'});` call from the Component you've navigated to? 

The only way I've seen is to access it via `this.props.navigation.getParams('param1')` but without the navigation prop, how do you call `getParams`?

**Comment by souzamarcos** [comment]
@erinha It's better use the params in the React Component whereas you could set and get the param1 value by redux if you want to access it outside the React Component.

**Comment by jprivillaso** [comment]
@erinha You can access the parameters the same way as the documentation states: 
`this.props.navigation.getParam('paramName', 'default-value-for-param')`
For more info, read the [official docs ](https://reactnavigation.org/docs/en/params.html)

**Comment by Burgeon8interactive** [comment]
@vinipachecov.. It works perfectly.. Thanks

**Comment by kodayashi** [comment]
Has anyone tried this approach in a redux-saga? Isn't working for me. In my saga, I call

```javascript
yield put(NavigationService.navigate('SomeRoute'))
```

And redux-saga throws an error:

```
"uncaught at check",
"put(action): argument action is undefined"
```

I've created the `NavigationService` per react-navigation's recommendation.

**Comment by pedrouid** [comment]
> Follow the tutorial by React Navigation (https://reactnavigation.org/docs/en/navigating-without-navigation-prop.html#docsNav)
> 
> "The presented approach is useful in situations when you want to trigger a navigation action from places where you do not have access to the navigation prop, or if you're looking for an alternative to using the navigation prop."
> 
> Create the NavigationService.js:
> 
> ```
> import { NavigationActions } from 'react-navigation';
> 
> let _navigator;
> 
> function setTopLevelNavigator(navigatorRef) {
>   _navigator = navigatorRef;
> }
> 
> function navigate(routeName, params) {
>   _navigator.dispatch(
>     NavigationActions.navigate({
>       routeName,
>       params,
>     })
>   );
> }
> 
> // add other navigation functions that you need and export them
> 
> export default {
>   navigate,
>   setTopLevelNavigator,
> };
> ```
> "You can get access to a navigator through a ref and pass it to the NavigationService which we will later use to navigate. Use this only with the top-level (root) navigator of your app."
> 
> App.js:
> 
> ```
> import NavigationService from './NavigationService';
> 
> const TopLevelNavigator = createStackNavigator({ /* ... */ })
> 
> class App extends React.Component {
>   // ...
> 
>   render() {
>     return (
>       <TopLevelNavigator
>         ref={navigatorRef => {
>           NavigationService.setTopLevelNavigator(navigatorRef);
>         }}
>       />
>     );
>   }
> }
> ```
> In you code import the NavigationService
> 
> ```
> import NavigationService from 'path-to-NavigationService.js';
> ```
> and navigate like this example:
> 
> ```
> NavigationService.navigate('Screen2', { param1: 'a', param2: 'b'  });
> ```

I've followed the tutorial and I've managed to navigate to the write screen but it's not passing`screenProps` properly.

```
const MainTabNavigator = createBottomTabNavigator({ ... })

const ModalStack = createStackNavigator(
  {
    Request: ModalRequest
  },
  {
    initialRouteName: "Request",
    headerMode: "none",
    mode: "modal"
  }
);


const AppNavigator = createStackNavigator(
  {
    Main: MainTabNavigator,
    Modal: ModalStack
  },
  {
    initialRouteName: "Main",
    headerMode: "none",
    mode: "modal"
  }
);

const AppContainer = createAppContainer(AppNavigator);

class App extends React.Component {
  render = () => (
    <AppContainer ref={navigatorRef => setTopLevelNavigator(navigatorRef)} />
  );
}
```

I first tried to do as the tutorial demonstrated to pass params as screenProps, but that didn't work then I saw a solution for passing props between nested navigators

```
function openRequestModal(params?: NavigationParams) {
  navigate(
    "Modal",
    {},
    {
      type: "Navigation/NAVIGATE",
      routeName: "Request",
      params
    }
  );
}
```

Unfortunately this doesn't seem to be the solution either. Any ideas?

**Comment by meisamnazari71** [comment]
The easiest way is using navigate as a props and connect it to your action:
Action.js file:
export const Action = (navigate) => {
    return (dispatch) => {
                navigate('Screen want to navigate' );
                dispatch({
                    type: ACTION,
                });
    }
};

Screen.js:

class Screen extends Component {

    onButtonPress=()=>{
        const navigate = this.props.navigation.navigate;
        this.props.Action(navigate)
    };

    render() {
        return (
            <View>
                <TouchableOpacity
                    style={{height: 30}}
                    onPress={this.onButtonPress}
                >
                    <Text>navigate to other page</Text>
                </TouchableOpacity>
            </View>
        );
    }
}

const mapStateToProps = state => {
........
    }
};

export default connect(mapStateToProps,{Action})(Screen);

**Comment by kelvinlemus** [comment]
Hi @meisamnazari71, I don't know if it's the best but I'm using in the same way

**Comment by ahmedelsayed11** [comment]
> From [daose/react-native-template#1](https://github.com/daose/react-native-template/issues/1), #153
> 
> You can `import { NavigationActions } from 'react-navigation'`, and then if you have `dispatch`, you can do
> 
> ```js
> dispatch(NavigationActions.navigate({ routeName: 'Route' });
> ```
> 
> If you use `redux-saga`:
> 
> ```js
> yield put(NavigationActions.navigate({ routeName: 'Route' }); // import put from redux-saga/effects
> ```
> 
> There's a lot more you can do with it as well, you can see the available actions at [NavigationActions.js](https://github.com/react-community/react-navigation/blob/master/src/NavigationActions.js)

i tried this in my action file but nothing happens please help

**Comment by igor-lemon** [comment]
The same issue. Via navigator `dispatch` method all work ok, but if replace to redux `dispatch` method it doesn't work.
I want to subscribe on `NAVIGATE` action but I can't do because the events system works very strange (imho).

**Comment by losheredos** [comment]
@ahmedelsayed11 I had same problem and used this documentation now its working!

check it: https://reactnavigation.org/docs/en/navigating-without-navigation-prop.html

**Comment by JorgeWSantos** [comment]
i solve the problem passing a callback function to function parameter and calling the function inside saga function

-------------------
action.js:

export function checkTokenDeviceRequest(
  id_onesignal,
  token,
  callbackFunctionOnSuccess = () => {},
  callbackFunctionOnFailure = () => {},
) {
  return {
    type: '@auth/CHECK_TOKEN_DEVICE_REQUEST',
    payload: {
      id_onesignal,
      token,
      callbackFunctionOnSuccess,
      callbackFunctionOnFailure,
    },
  };
}

--------------------
sagas.js:

export function* checkTokenDevice({ payload }) {
  const { token, callbackFunctionOnSuccess, callbackFunctionOnFailure } =
    payload;

 callbackFunctionOnSuccess();
....
...
}
